README.md
# disksinfo

[![travis Status](https://travis-ci.org/mh-cbon/disksinfo.svg?branch=master)](https://travis-ci.org/mh-cbon/disksinfo)[![appveyor Status](https://ci.appveyor.com/api/projects/status/github/mh-cbon/disksinfo?branch=master&svg=true)](https://ci.appveyor.com/project/mh-cbon/disksinfo)
[![Go Report Card](https://goreportcard.com/badge/github.com/mh-cbon/disksinfo)](https://goreportcard.com/report/github.com/mh-cbon/disksinfo)

[![GoDoc](https://godoc.org/github.com/mh-cbon/disksinfo?status.svg)](http://godoc.org/github.com/mh-cbon/disksinfo)


Package disksinfo provides the list of partitions


Compatible __windows__ / __linux__

# Install

#### Go

```sh
go get github.com/mh-cbon/disksinfo
```


# Usage


###### $ disksinfo 
```sh
[
    {
        "Label": "",
        "IsRemovable": false,
        "Size": "1,9G",
        "SpaceLeft": "1,9G",
        "Path": "devtmpfs",
        "MountPath": "/dev"
    },
    {
        "Label": "",
        "IsRemovable": false,
        "Size": "1,9G",
        "SpaceLeft": "1,9G",
        "Path": "tmpfs",
        "MountPath": "/dev/shm"
    },
    {
        "Label": "",
        "IsRemovable": false,
        "Size": "32G",
        "SpaceLeft": "3,4G",
        "Path": "/dev/mapper/fedora-root",
        "MountPath": "/"
    },
    {
        "Label": "",
        "IsRemovable": false,
        "Size": "126G",
        "SpaceLeft": "15G",
        "Path": "/dev/sda5",
        "MountPath": "/home"
    },
    {
        "Label": "whatever",
        "IsRemovable": true,
        "Size": "932G",
        "SpaceLeft": "515G",
        "Path": "/dev/sdb1",
        "MountPath": "/run/media/mh-cbon/whatever"
    },
    {
        "Label": "Recovery",
        "IsRemovable": false,
        "Size": "",
        "SpaceLeft": "",
        "Path": "/dev/sda1",
        "MountPath": ""
    },
    {
        "Label": "stockage",
        "IsRemovable": false,
        "Size": "",
        "SpaceLeft": "",
        "Path": "/dev/sda6",
        "MountPath": ""
    },
    {
        "Label": "System Reserved",
        "IsRemovable": false,
        "Size": "",
        "SpaceLeft": "",
        "Path": "/dev/sda2",
        "MountPath": ""
    }
]
```

# API example


###### > main.go
```go
// Package disksinfo provides the list of partitions
package main

import (
	"encoding/json"
	"os"

	"github.com/mh-cbon/disksinfo/diskinfo"
)

func main() {
	loader := diskinfo.NewMultiOsLoader()
	p, err := loader.Load()
	if err != nil {
		panic(err)
	}
	enc := json.NewEncoder(os.Stdout)
	enc.SetIndent("", "    ")
	if err := enc.Encode(p); err != nil {
		panic(err)
	}
}
```

# Recipes

#### Release the project

```sh
gump patch -d # check
gump patch # bump
```


CHANGELOG.md
# Changelog - disksinfo

### 0.0.1

__Changes__

- init

__Contributors__

- mh-cbon

Released by mh-cbon, Thu 16 Mar 2017 -
[see the diff](https://github.com/mh-cbon/disksinfo/compare/2ac728a6aa4371d2d8665b836e4f0b21fdf73216...0.0.1#diff)
______________




diskinfo/all_test.go
package diskinfo

type parseTable struct {
	in        string
	expectErr error
	expectOut []*Properties
	path      string // for ls test
}


diskinfo/linux_test.go
// +build linux

package diskinfo

import (
	"bufio"
	"bytes"
	"testing"
)

func TestDfParser(t *testing.T) {

	testsTable := []parseTable{
		parseTable{
			in: `Sys. de fichiers        Taille Utilisé Dispo Uti% Monté sur
devtmpfs                  1,9G       0  1,9G   0% /dev
tmpfs                     1,9G     67M  1,9G   4% /dev/shm
tmpfs                     1,9G    1,6M  1,9G   1% /run
tmpfs                     1,9G       0  1,9G   0% /sys/fs/cgroup
/dev/mapper/fedora-root    32G     26G  4,1G  87% /
`,
			expectErr: nil,
			expectOut: []*Properties{
				&Properties{
					Size:      "1,9G",
					SpaceLeft: "1,9G",
					Path:      "devtmpfs",
					MountPath: "/dev",
				},
				&Properties{
					Size:      "1,9G",
					SpaceLeft: "1,9G",
					Path:      "tmpfs",
					MountPath: "/dev/shm",
				},
				&Properties{
					Size:      "1,9G",
					SpaceLeft: "1,9G",
					Path:      "tmpfs",
					MountPath: "/run",
				},
				&Properties{
					Size:      "1,9G",
					SpaceLeft: "1,9G",
					Path:      "tmpfs",
					MountPath: "/sys/fs/cgroup",
				},
				&Properties{
					Size:      "32G",
					SpaceLeft: "4,1G",
					Path:      "/dev/mapper/fedora-root",
					MountPath: "/",
				},
				&Properties{
					Size:      "32G",
					SpaceLeft: "4,1G",
					Path:      "/dev/mapper/fedora-root",
					MountPath: "/",
				},
			},
		},
	}

	for i, testTable := range testsTable {

		var b bytes.Buffer
		r := NewDfReader(bufio.NewReader(&b))
		b.WriteString(testTable.in)

		res, err := r.Read()
		if err != nil && testTable.expectErr != err {
			t.Fatalf("Test(%v): Unexpected error %v", i, err)
		}

		for _, p := range res {
			found := PropertiesList(testTable.expectOut).FindByPath(p.Path)
			if found == nil {
				t.Errorf("Test(%v): Unexpected property %q not found\n%#v\ntestTable.in=\n%v", i, p.Path, p, testTable.in)
			}
		}

		for _, p := range testTable.expectOut {
			found := PropertiesList(res).FindByPath(p.Path)
			if found == nil {
				t.Errorf("Test(%v): Property %q not found\n%#v\ntestTable.in=\n%v", i, p.Path, p, testTable.in)
			}
		}
	}
}

func TestLsParser(t *testing.T) {

	testsTable := []parseTable{
		parseTable{
			in: `total 0
lrwxrwxrwx 1 root root 10 27 févr. 11:04 Recovery -> ../../sda1
lrwxrwxrwx 1 root root 10 27 févr. 11:04 stockage -> ../../sda6
lrwxrwxrwx 1 root root 10 27 févr. 11:04 'System\x20Reserved' -> ../../sda2
`,
			path:      "/dev/disk/by-label/",
			expectErr: nil,
			expectOut: []*Properties{
				&Properties{
					Label: "Recovery",
					Path:  "/dev/sda1",
				},
				&Properties{
					Label: "stockage",
					Path:  "/dev/sda6",
				},
				&Properties{
					Label: "System Reserved",
					Path:  "/dev/sda2",
				},
			},
		},
	}

	for i, testTable := range testsTable {

		var b bytes.Buffer
		r := NewLsReader(bufio.NewReader(&b))
		b.WriteString(testTable.in)

		res, err := r.Read(testTable.path)
		if err != nil && testTable.expectErr != err {
			t.Fatalf("Test(%v): Unexpected error %v", i, err)
		}

		for _, p := range res {
			found := PropertiesList(testTable.expectOut).FindByPath(p.Path)
			if found == nil {
				t.Errorf("Test(%v): Unexpected property %q not found\n%#v\ntestTable.in=\n%v", i, p.Path, p, testTable.in)
			}
		}

		for _, p := range testTable.expectOut {
			found := PropertiesList(res).FindByPath(p.Path)
			if found == nil {
				t.Errorf("Test(%v): Property %q not found\n%#v\ntestTable.in=\n%v", i, p.Path, p, testTable.in)
			}
		}
	}
}

func TestMountParser(t *testing.T) {

	testsTable := []parseTable{
		parseTable{
			in: `
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,relatime)
proc on /proc type proc (rw,nosuid,nodev,noexec,relatime)
devtmpfs on /dev type devtmpfs (rw,nosuid,size=1966944k,nr_inodes=491736,mode=755)
securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000)
tmpfs on /run type tmpfs (rw,nosuid,nodev,mode=755)
tmpfs on /sys/fs/cgroup type tmpfs (ro,nosuid,nodev,noexec,mode=755)
cgroup on /sys/fs/cgroup/systemd type cgroup (rw,nosuid,nodev,noexec,relatime,xattr,release_agent=/usr/lib/systemd/systemd-cgroups-agent,name=systemd)
pstore on /sys/fs/pstore type pstore (rw,nosuid,nodev,noexec,relatime)
cgroup on /sys/fs/cgroup/net_cls,net_prio type cgroup (rw,nosuid,nodev,noexec,relatime,net_cls,net_prio)
cgroup on /sys/fs/cgroup/freezer type cgroup (rw,nosuid,nodev,noexec,relatime,freezer)
cgroup on /sys/fs/cgroup/hugetlb type cgroup (rw,nosuid,nodev,noexec,relatime,hugetlb)
cgroup on /sys/fs/cgroup/devices type cgroup (rw,nosuid,nodev,noexec,relatime,devices)
cgroup on /sys/fs/cgroup/cpu,cpuacct type cgroup (rw,nosuid,nodev,noexec,relatime,cpu,cpuacct)
cgroup on /sys/fs/cgroup/cpuset type cgroup (rw,nosuid,nodev,noexec,relatime,cpuset)
cgroup on /sys/fs/cgroup/perf_event type cgroup (rw,nosuid,nodev,noexec,relatime,perf_event)
cgroup on /sys/fs/cgroup/memory type cgroup (rw,nosuid,nodev,noexec,relatime,memory)
cgroup on /sys/fs/cgroup/blkio type cgroup (rw,nosuid,nodev,noexec,relatime,blkio)
cgroup on /sys/fs/cgroup/pids type cgroup (rw,nosuid,nodev,noexec,relatime,pids)
configfs on /sys/kernel/config type configfs (rw,relatime)
/dev/mapper/fedora-root on / type ext4 (rw,relatime,data=ordered)
systemd-1 on /proc/sys/fs/binfmt_misc type autofs (rw,relatime,fd=42,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=14678)
mqueue on /dev/mqueue type mqueue (rw,relatime)
hugetlbfs on /dev/hugepages type hugetlbfs (rw,relatime)
debugfs on /sys/kernel/debug type debugfs (rw,relatime)
nfsd on /proc/fs/nfsd type nfsd (rw,relatime)
tmpfs on /tmp type tmpfs (rw,nosuid,nodev)
/dev/sda5 on /home type ext4 (rw,relatime,data=ordered)
sunrpc on /var/lib/nfs/rpc_pipefs type rpc_pipefs (rw,relatime)
binfmt_misc on /proc/sys/fs/binfmt_misc type binfmt_misc (rw,relatime)
tmpfs on /run/user/1001 type tmpfs (rw,nosuid,nodev,relatime,size=396012k,mode=700,uid=1001,gid=1001)
fusectl on /sys/fs/fuse/connections type fusectl (rw,relatime)
gvfsd-fuse on /run/user/1001/gvfs type fuse.gvfsd-fuse (rw,nosuid,nodev,relatime,user_id=1001,group_id=1001)
/dev/sdb1 on /run/media/mh-cbon/whatever type fuseblk (rw,nosuid,nodev,relatime,user_id=0,group_id=0,default_permissions,allow_other,blksize=4096,uhelper=udisks2) [whatever]
`,
			expectErr: nil,
			expectOut: []*Properties{
				&Properties{
					MountPath: "/",
					Path:      "/dev/mapper/fedora-root",
				},
				&Properties{
					MountPath: "/home",
					Path:      "/dev/sda5",
				},
				&Properties{
					Label:     "Whatever",
					MountPath: "/run/media/mh-cbon/whatever",
					Path:      "/dev/sdb1",
				},
			},
		},
	}

	for i, testTable := range testsTable {

		var b bytes.Buffer
		r := NewMountReader(bufio.NewReader(&b))
		b.WriteString(testTable.in)

		res, err := r.Read()
		if err != nil && testTable.expectErr != err {
			t.Fatalf("Test(%v): Unexpected error %v", i, err)
		}

		for _, p := range res {
			found := PropertiesList(testTable.expectOut).FindByPath(p.Path)
			if found == nil {
				t.Errorf("Test(%v): Unexpected property %q not found\n%#v\ntestTable.in=\n%v", i, p.Path, p, testTable.in)
			}
		}

		for _, p := range testTable.expectOut {
			found := PropertiesList(res).FindByPath(p.Path)
			if found == nil {
				t.Errorf("Test(%v): Property %q not found\n%#v\ntestTable.in=\n%v", i, p.Path, p, testTable.in)
			}
		}
	}
}


diskinfo/windows_test.go
package diskinfo

import (
	"bufio"
	"bytes"
	"testing"
)

func TestWmicParser(t *testing.T) {

	testsTable := []parseTable{
		parseTable{
			//todo need to fix this output with a real one from windows.
			in: `
Caption  Description       FileSystem
C:       Local Fixed Disk  NTFS
`,
			expectErr: nil,
			expectOut: []*Properties{
				&Properties{
					Path:      "C:",
					MountPath: "C:",
				},
			},
		},
	}

	for i, testTable := range testsTable {

		var b bytes.Buffer
		r := NewWmicReader(bufio.NewReader(&b))
		b.WriteString(testTable.in)

		res, err := r.Read()
		if err != nil && testTable.expectErr != err {
			t.Fatalf("Test(%v): Unexpected error %v", i, err)
		}

		for _, p := range res {
			found := PropertiesList(testTable.expectOut).FindByPath(p.Path)
			if found == nil {
				t.Errorf("Test(%v): Unexpected property %q not found\n%#v\ntestTable.in=\n%v", i, p.Path, p, testTable.in)
			}
		}

		for _, p := range testTable.expectOut {
			found := PropertiesList(res).FindByPath(p.Path)
			if found == nil {
				t.Errorf("Test(%v): Property %q not found\n%#v\ntestTable.in=\n%v", i, p.Path, p, testTable.in)
			}
		}
	}
}


diskinfo/lib.go
// Package diskinfo provides the list of partitions
package diskinfo

import "runtime"

// PropertiesLoader can load the list of partitions and their properties.
type PropertiesLoader interface {
	Load() ([]*Properties, error)
}

// NewMultiOsLoader prepares a multi os loader for the current runtime operating system.
func NewMultiOsLoader() PropertiesLoader {
	var loader PropertiesLoader
	loader = &LinuxLoader{}
	if runtime.GOOS == "windows" {
		loader = &WindowsLoader{}
	}
	return loader
}

// Properties provides information about a partition on the system.
type Properties struct {
	Label       string
	IsRemovable bool
	Size        string
	SpaceLeft   string
	Path        string
	MountPath   string
}

// NewProperties is a constructor.
func NewProperties() *Properties {
	return &Properties{}
}

// PropertiesList is type alias to []*Properties
type PropertiesList []*Properties

// Merge some []*Properties into this list. what is a property name of Properties.
// Partitions are matched by their Path value.
func (l PropertiesList) Merge(some PropertiesList, what ...string) []*Properties {
	for i, d := range l {
		s := some.FindByPath(d.Path)
		if s != nil {
			for _, w := range what {
				switch w {
				case "Label":
					if s.Label != "" {
						d.Label = s.Label
					}
				case "Size":
					if s.Size != "" {
						d.Size = s.Size
					}
				case "SpaceLeft":
					if s.SpaceLeft != "" {
						d.SpaceLeft = s.SpaceLeft
					}
				case "MountPath":
					if s.MountPath != "" {
						d.MountPath = s.MountPath
					}
				case "Path":
					if s.Path != "" {
						d.Path = s.Path
					}
				case "IsRemovable":
					d.IsRemovable = s.IsRemovable
				}
			}
			l[i] = d
		}
	}
	return l
}

// Append some []*Properties into this list. what is a property name of Properties.
// Partitions are matched by their Path value.
func (l PropertiesList) Append(some PropertiesList) []*Properties {
	for _, d := range some {
		s := l.FindByPath(d.Path)
		if s == nil {
			l = append(l, d)
		}
	}
	return l
}

// FindByPath search a partition by its path.
// Partitions are matched by their Path value.
func (l PropertiesList) FindByPath(path string) *Properties {
	for _, d := range l {
		if d.Path == path {
			return d
		}
	}
	return nil
}


diskinfo/windows.go
package diskinfo

import (
	"io"
	"os"
	"os/exec"
	"strings"
)

// WindowsLoader can load disk information for a windows system using wmmic.
// Its compatile with windows 7+.
type WindowsLoader struct {
}

// Load queries wmmic and parses its result to return the list of partitions with their properties.
func (w *WindowsLoader) Load() ([]*Properties, error) {
	p := []*Properties{}
	return p, nil
}

func runWmic() ([]*Properties, error) {
	var ret []*Properties
	cmd := exec.Command("wmic", "logicaldisk", "get", "caption,description,name,freespace,size")
	cmd.Stderr = os.Stderr

	sink, err := cmd.StdoutPipe()
	if err != nil {
		return ret, err
	}

	if err2 := cmd.Start(); err2 != nil {
		return ret, err2
	}

	ret, err = NewWmicReader(sink).Read()
	if err != nil {
		return ret, err
	}

	if err2 := cmd.Wait(); err2 != nil {
		return ret, err2
	}

	return ret, err
}

// WmicReader reads a wmic logicaldisk command output.
type WmicReader struct {
	r    io.Reader
	line string
}

// NewWmicReader parses a wmic logicaldisk command output.
func NewWmicReader(r io.Reader) *WmicReader {
	return &WmicReader{r: r}
}

// Read parses a wmic logicaldisk command output, it returns a list of properies for each property found.
func (l *WmicReader) Read() ([]*Properties, error) {

	var ret []*Properties
	/*
	   Caption  Description       FileSystem
	   C:       Local Fixed Disk  NTFS
	   D:       CD-ROM Disc
	   E:       CD-ROM Disc
	   F:       Removable Disk
	   G:       Removable Disk
	   H:       Removable Disk
	   I:       Removable Disk
	   J:       Removable Disk    FAT
	*/
	/*
	     write-host "Access: " $objItem.Access
	   write-host "Availability: " $objItem.Availability
	   write-host "Block Size: " $objItem.BlockSize
	   write-host "Caption: " $objItem.Caption
	   write-host "Compressed: " $objItem.Compressed
	   write-host "Configuration Manager Error Code: " $objItem.ConfigManagerErrorCode
	   write-host "Configuration Manager User Configuration: " $objItem.ConfigManagerUserConfig
	   write-host "Creation Class Name: " $objItem.CreationClassName
	   write-host "Description: " $objItem.Description
	   write-host "Device ID: " $objItem.DeviceID
	   write-host "Drive Type: " $objItem.DriveType
	   write-host "Error Cleared: " $objItem.ErrorCleared
	   write-host "Error Description: " $objItem.ErrorDescription
	   write-host "Error Methodology: " $objItem.ErrorMethodology
	   write-host "File System: " $objItem.FileSystem
	   write-host "Free Space: " $objItem.FreeSpace
	   write-host "Installation Date: " $objItem.InstallDate
	   write-host "Last Error Code: " $objItem.LastErrorCode
	   write-host "Maximum Component Length: " $objItem.MaximumComponentLength
	   write-host "Media Type: " $objItem.MediaType
	   write-host "Name: " $objItem.Name
	   write-host "Number Of Blocks: " $objItem.NumberOfBlocks
	   write-host "PNP Device ID: " $objItem.PNPDeviceID
	   write-host "Power Management Capabilities: " $objItem.PowerManagementCapabilities
	   write-host "Power Management Supported: " $objItem.PowerManagementSupported
	   write-host "Provider Name: " $objItem.ProviderName
	   write-host "Purpose: " $objItem.Purpose
	   write-host "Quotas Disabled: " $objItem.QuotasDisabled
	   write-host "Quotas Incomplete: " $objItem.QuotasIncomplete
	   write-host "Quotas Rebuilding: " $objItem.QuotasRebuilding
	   write-host "Size: " $objItem.Size
	   write-host "Status: " $objItem.Status
	   write-host "Status Information: " $objItem.StatusInfo
	   write-host "Supports Disk Quotas: " $objItem.SupportsDiskQuotas
	   write-host "Supports File-Based Compression: " $objItem.SupportsFileBasedCompression
	   write-host "System Creation Class Name: " $objItem.SystemCreationClassName
	   write-host "System Name: " $objItem.SystemName
	   write-host "Volume Dirty: " $objItem.VolumeDirty
	   write-host "Volume Name: " $objItem.VolumeName
	   write-host "Volume Serial Number: " $objItem.VolumeSerialNumber
	*/
	i := 0
	headersLen := []int{}

	b := NewLineReader(l.r)
	var err error
	for {
		line, err2 := b.ReadLine()
		err = err2

		if line != "" {

			if i == 0 {
				headers := strings.Split(line, "")
				last := ""
				l := 0
				for _, p := range headers {
					if p != " " {
						if last == " " {
							headersLen = append(headersLen, l)
						}
					}
					l++
					last = p
				}
				headersLen = append(headersLen, l)

			} else {
				c := 0
				s := []string{}
				for e, l := range headersLen {
					if e == 0 {
						s = append(s, line[:l])
					} else if e == len(headersLen)-1 {
						s = append(s, line[c:])
					} else {
						s = append(s, line[c:l])
					}
					c = l
				}
				p := NewProperties()
				p.Path = strings.TrimSpace(s[0])
				s[1] = strings.TrimSpace(s[1])
				p.IsRemovable = s[1] == "Removable Disk"
				ret = append(ret, p)
			}
			i++
		}

		if err != nil {
			break
		}
	}

	if err == io.EOF {
		err = nil
	}

	return ret, err
}


diskinfo/linux.go
package diskinfo

import (
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
)

// LinuxLoader can load disk information for a linux system using df / fidsk.
type LinuxLoader struct {
}

// Load returns the list of partition found and their properties.
func (l *LinuxLoader) Load() ([]*Properties, error) {
	//-
	var ret PropertiesList

	if temp, err := runDf(); err != nil {
		return ret, err
	} else {
		ret = ret.Append(temp)
	}
	//-
	if temp, err := runLsLabel(); err != nil {
		return ret, err
	} else {
		ret = ret.Append(temp)
	}
	//-
	if temp, err := runLsUsb(); err != nil {
		return ret, err
	} else {
		ret = ret.Merge(temp, "IsRemovable")
	}
	//-
	if temp, err := runMount(); err != nil {
		return ret, err
	} else {
		ret = ret.Merge(temp, "Label")
	}
	//-
	return ret, nil
}

func runLsLabel() ([]*Properties, error) {
	var ret []*Properties

	path := "/dev/disk/by-label/"
	disks, err := runLs(path)
	if err != nil {
		return disks, err
	}

	for _, disk := range disks {
		ret = append(ret, disk)
	}

	return ret, nil
}

func runLsUsb() ([]*Properties, error) {
	var ret []*Properties

	path := "/dev/disk/by-id/"
	disks, err := runLs(path)
	if err != nil {
		return disks, err
	}

	for _, disk := range disks {
		if strings.HasPrefix(disk.Label, "usb-") {
			disk.IsRemovable = true
			ret = append(ret, disk)
		}
	}

	return ret, nil
}

// LsReader ...
type LsReader struct {
	r    io.Reader
	line string
}

// NewLsReader ...
func NewLsReader(r io.Reader) *LsReader {
	return &LsReader{r: r}
}

// Read ...
func (l *LsReader) Read(path string) ([]*Properties, error) {

	/*
			   total 0
		     lrwxrwxrwx 1 root root  9 27 févr. 13:21 /dev/disk/by-id/usb-TOSHIBA_External_USB_3.0_20141114026944-0:0 -> ../../sdb
		     lrwxrwxrwx 1 root root 10 27 févr. 13:21 /dev/disk/by-id/usb-TOSHIBA_External_USB_3.0_20141114026944-0:0-part1 -> ../../sdb1
	*/

	var ret []*Properties
	i := 0

	b := NewLineReader(l.r)
	var err error
	for {
		line, err2 := b.ReadLine()
		err = err2

		if i > 0 && line != "" {
			props := strings.Split(line, " ")
			s := []string{}
			for _, p := range props {
				p = strings.TrimSpace(p)
				if p != "" {
					s = append(s, p)
				}
			}
			p := NewProperties()
			name := s[8]
			if strings.HasPrefix(name, "'") {
				name = name[1:]
			}
			if strings.HasSuffix(name, "'") {
				name = name[:len(name)-1]
			}
			p.Label = strings.Replace(name, "\\x20", " ", -1)
			abs, err2 := filepath.Abs(filepath.Join(path, s[10]))
			if err2 != nil {
				panic(err2)
			}
			p.Path = abs
			ret = append(ret, p)
		}

		if err != nil {
			break
		}
		i++
	}

	if err == io.EOF {
		err = nil
	}

	return ret, err
}

func runLs(path string) ([]*Properties, error) {
	var ret []*Properties
	_, err := os.Stat(path)
	// when there is not any usb disk connected on the computer
	// skip err and return to avoid errors about
	// missing directory.
	if os.IsNotExist(err) {
		return ret, nil
	}
	cmd := exec.Command("ls", "-l", path)
	cmd.Stderr = os.Stderr

	sink, err := cmd.StdoutPipe()
	if err != nil {
		return ret, err
	}

	if err2 := cmd.Start(); err2 != nil {
		return ret, err2
	}

	ret, err = NewLsReader(sink).Read(path)
	if err != nil {
		return ret, err
	}

	if err2 := cmd.Wait(); err2 != nil {
		return ret, err2
	}

	return ret, err
}

func runDf() ([]*Properties, error) {
	var ret []*Properties
	cmd := exec.Command("df", "-h")
	cmd.Stderr = os.Stderr

	sink, err := cmd.StdoutPipe()
	if err != nil {
		return ret, err
	}

	if err2 := cmd.Start(); err2 != nil {
		return ret, err2
	}

	ret, err = NewDfReader(sink).Read()
	if err != nil {
		return ret, err
	}

	if err2 := cmd.Wait(); err2 != nil {
		return ret, err2
	}

	return ret, err
}

// DfReader ...
type DfReader struct {
	r    io.Reader
	line string
}

// NewDfReader ...
func NewDfReader(r io.Reader) *DfReader {
	return &DfReader{r: r}
}

// Read ...
func (l *DfReader) Read() ([]*Properties, error) {

	/*
	   Sys. de fichiers        Taille Utilisé Dispo Uti% Monté sur
	   devtmpfs                  1,9G       0  1,9G   0% /dev
	   tmpfs                     1,9G     54M  1,9G   3% /dev/shm
	*/

	var ret []*Properties
	i := 0

	b := NewLineReader(l.r)
	var err error
	for {
		line, err2 := b.ReadLine()
		err = err2

		if i > 0 && line != "" {
			props := strings.Split(line, " ")
			s := []string{}
			for _, p := range props {
				p = strings.TrimSpace(p)
				if p != "" {
					s = append(s, p)
				}
			}
			p := NewProperties()
			p.Size = s[1]
			p.SpaceLeft = s[3]
			p.Path = s[0]
			p.MountPath = s[5]
			ret = append(ret, p)
		}

		if err != nil {
			break
		}
		i++
	}

	if err == io.EOF {
		err = nil
	}
	return ret, err
}

func runMount() ([]*Properties, error) {
	var ret []*Properties
	cmd := exec.Command("mount", "-l")
	cmd.Stderr = os.Stderr

	sink, err := cmd.StdoutPipe()
	if err != nil {
		return ret, err
	}

	if err2 := cmd.Start(); err2 != nil {
		return ret, err2
	}

	ret, err = NewMountReader(sink).Read()
	if err != nil {
		return ret, err
	}

	if err2 := cmd.Wait(); err2 != nil {
		return ret, err2
	}

	return ret, err
}

// MountReader ...
type MountReader struct {
	r    io.Reader
	line string
}

// NewMountReader ...
func NewMountReader(r io.Reader) *MountReader {
	return &MountReader{r: r}
}

var lineR = regexp.MustCompile(`\s*([^\s]+)\s+on\s+([^\s]+)\s+type\s+([^\s]+)\s+\(([^)]+)\)(\s+\[[^]]+\])?`)

// Read ...
func (l *MountReader) Read() ([]*Properties, error) {

	/*
		sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,relatime)
		proc on /proc type proc (rw,nosuid,nodev,noexec,relatime)
		devtmpfs on /dev type devtmpfs (rw,nosuid,size=1966944k,nr_inodes=491736,mode=755)
		securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)
		tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev)
		devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000)
		tmpfs on /run type tmpfs (rw,nosuid,nodev,mode=755)
		tmpfs on /sys/fs/cgroup type tmpfs (ro,nosuid,nodev,noexec,mode=755)
		cgroup on /sys/fs/cgroup/systemd type cgroup (rw,nosuid,nodev,noexec,relatime,xattr,release_agent=/usr/lib/systemd/systemd-cgroups-agent,name=systemd)
		pstore on /sys/fs/pstore type pstore (rw,nosuid,nodev,noexec,relatime)
		cgroup on /sys/fs/cgroup/net_cls,net_prio type cgroup (rw,nosuid,nodev,noexec,relatime,net_cls,net_prio)
		cgroup on /sys/fs/cgroup/freezer type cgroup (rw,nosuid,nodev,noexec,relatime,freezer)
		cgroup on /sys/fs/cgroup/hugetlb type cgroup (rw,nosuid,nodev,noexec,relatime,hugetlb)
		cgroup on /sys/fs/cgroup/devices type cgroup (rw,nosuid,nodev,noexec,relatime,devices)
		cgroup on /sys/fs/cgroup/cpu,cpuacct type cgroup (rw,nosuid,nodev,noexec,relatime,cpu,cpuacct)
		cgroup on /sys/fs/cgroup/cpuset type cgroup (rw,nosuid,nodev,noexec,relatime,cpuset)
		cgroup on /sys/fs/cgroup/perf_event type cgroup (rw,nosuid,nodev,noexec,relatime,perf_event)
		cgroup on /sys/fs/cgroup/memory type cgroup (rw,nosuid,nodev,noexec,relatime,memory)
		cgroup on /sys/fs/cgroup/blkio type cgroup (rw,nosuid,nodev,noexec,relatime,blkio)
		cgroup on /sys/fs/cgroup/pids type cgroup (rw,nosuid,nodev,noexec,relatime,pids)
		configfs on /sys/kernel/config type configfs (rw,relatime)
		/dev/mapper/fedora-root on / type ext4 (rw,relatime,data=ordered)
		systemd-1 on /proc/sys/fs/binfmt_misc type autofs (rw,relatime,fd=42,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=14678)
		mqueue on /dev/mqueue type mqueue (rw,relatime)
		hugetlbfs on /dev/hugepages type hugetlbfs (rw,relatime)
		debugfs on /sys/kernel/debug type debugfs (rw,relatime)
		nfsd on /proc/fs/nfsd type nfsd (rw,relatime)
		tmpfs on /tmp type tmpfs (rw,nosuid,nodev)
		/dev/sda5 on /home type ext4 (rw,relatime,data=ordered)
		sunrpc on /var/lib/nfs/rpc_pipefs type rpc_pipefs (rw,relatime)
		binfmt_misc on /proc/sys/fs/binfmt_misc type binfmt_misc (rw,relatime)
		tmpfs on /run/user/1001 type tmpfs (rw,nosuid,nodev,relatime,size=396012k,mode=700,uid=1001,gid=1001)
		fusectl on /sys/fs/fuse/connections type fusectl (rw,relatime)
		gvfsd-fuse on /run/user/1001/gvfs type fuse.gvfsd-fuse (rw,nosuid,nodev,relatime,user_id=1001,group_id=1001)
		/dev/sdb1 on /run/media/mh-cbon/whatever type fuseblk (rw,nosuid,nodev,relatime,user_id=0,group_id=0,default_permissions,allow_other,blksize=4096,uhelper=udisks2) [whatever]
	*/

	var ret []*Properties
	i := 0

	b := NewLineReader(l.r)
	var err error
	for {
		line, err2 := b.ReadLine()
		err = err2

		if line != "" {
			x := lineR.FindAllStringSubmatch(line, -1)
			if len(x) > 0 {
				s := x[0][1:]
				if s[0][:1] == "/" {
					p := NewProperties()
					p.MountPath = s[1]
					p.Path = s[0]
					if len(s) > 3 && len(s[4]) > 0 {
						p.Label = strings.TrimSpace(s[4])
						p.Label = p.Label[1 : len(p.Label)-1]
					}
					ret = append(ret, p)
				}
			}
		}

		if err != nil {
			break
		}
		i++
	}

	if err == io.EOF {
		err = nil
	}
	return ret, err
}
